consensus/istanbul/backend/announce.go

```diff
@@ -1161,10 +1180,59 @@ func (sb *Backend) generateEnodeCertificateMsg(version uint) (*istanbul.Message,
 	if err := msg.Sign(sb.Sign); err != nil {
 		return nil, err
 	}
-	logger.Trace("Generated Istanbul Enode Certificate message", "enodeCertificate", enodeCertificate, "address", msg.Address)
+
 	return msg, nil
 }
 
+// generateEnodeCertificateMsg generates an enode certificate message with the enode
+// this node is publicly accessible at. If this node is proxied, the proxy's
+// public enode is used.
+func (sb *Backend) generateEnodeCertificateMsgs(version uint) (map[enode.ID]*istanbul.Message, error) {
+	logger := sb.logger.New("func", "generateEnodeCertificateMsg")
+
+	externalEnodes := make(map[enode.ID]*enode.Node)
+	if sb.IsProxiedValidator() {
+		if valProxyAssignments, err := sb.proxyEngine.GetValidatorProxyAssignments(); err != nil {
+			return nil, err
+		} else {
+			for _, proxyNode := range valProxyAssignments {
+				if proxyNode != nil {
+					externalEnodes[proxyNode.ID()] = proxyNode
+				}
+			}
+		}
+	} else {
+		selfEnode := sb.p2pserver.Self()
+		externalEnodes[selfEnode.ID()] = selfEnode
+	}
+
+	enodeCertificateMsgs := make(map[enode.ID]*istanbul.Message)
+	for externalNodeID, externalNode := range externalEnodes {
+		enodeCertificate := &istanbul.EnodeCertificate{
+			EnodeURL: externalNode.URLv4(),
+			Version:  version,
+		}
+		enodeCertificateBytes, err := rlp.EncodeToBytes(enodeCertificate)
+		if err != nil {
+			return nil, err
+		}
+		msg := &istanbul.Message{
+			Code:    istanbul.EnodeCertificateMsg,
+			Address: sb.Address(),
+			Msg:     enodeCertificateBytes,
+		}
+		// Sign the message
+		if err := msg.Sign(sb.Sign); err != nil {
+			return nil, err
+		}
+
+		enodeCertificateMsgs[externalNodeID] = msg
+	}
+
+	logger.Trace("Generated Istanbul Enode Certificate messages", "enodeCertificateMsgs", enodeCertificateMsgs)
+	return enodeCertificateMsgs, nil
+}
+
 // handleEnodeCertificateMsg handles an enode certificate message.
 // If this node is a proxy and the enode certificate is from a remote validator
 // (ie not the proxied validator), this node will forward the enode certificate
```

Comment:
This does not seem to be true anymore. It seems to me that if this function is called on a proxy, it will validate the certificate and do nothing (?)